File: src/meta_schedule/builder/builder.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include <tvm/ffi/reflection/registry.h>

#include "../utils.h"

namespace tvm {
namespace meta_schedule {

/******** Constructors ********/

BuilderInput::BuilderInput(IRModule mod, Target target,
                           ffi::Optional<ffi::Map<ffi::String, runtime::Tensor>> params) {
  ObjectPtr<BuilderInputNode> n = ffi::make_object<BuilderInputNode>();
  n->mod = std::move(mod);
  n->target = std::move(target);
  n->params = std::move(params);
  data_ = std::move(n);
}

BuilderResult::BuilderResult(ffi::Optional<ffi::String> artifact_path,
                             ffi::Optional<ffi::String> error_msg) {
  ObjectPtr<BuilderResultNode> n = ffi::make_object<BuilderResultNode>();
  n->artifact_path = std::move(artifact_path);
  n->error_msg = std::move(error_msg);
  data_ = std::move(n);
}

Builder Builder::PyBuilder(BuilderNode::FBuild f_build) {
  ObjectPtr<PyBuilderNode> n = ffi::make_object<PyBuilderNode>();
  n->f_build = std::move(f_build);
  return Builder(std::move(n));
}

/******** FFI ********/

TVM_FFI_STATIC_INIT_BLOCK() {
  BuilderInputNode::RegisterReflection();
  BuilderResultNode::RegisterReflection();
  PyBuilderNode::RegisterReflection();
}

TVM_FFI_STATIC_INIT_BLOCK() {
  namespace refl = tvm::ffi::reflection;
  refl::GlobalDef()
      .def("meta_schedule.BuilderInput",
           [](IRModule mod, Target target,
              ffi::Optional<ffi::Map<ffi::String, runtime::Tensor>> params) -> BuilderInput {
             return BuilderInput(mod, target, params);
           })
      .def("meta_schedule.BuilderResult",
           [](ffi::Optional<ffi::String> artifact_path, ffi::Optional<ffi::String> error_msg)
               -> BuilderResult { return BuilderResult(artifact_path, error_msg); })
      .def_method("meta_schedule.BuilderBuild", &BuilderNode::Build)
      .def("meta_schedule.BuilderPyBuilder", Builder::PyBuilder);
}

}  // namespace meta_schedule
}  // namespace tvm

File: python/tvm/s_tir/meta_schedule/builder/builder.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""Meta Schedule builders that translate IRModule to runtime.Module, and then export"""

from __future__ import annotations

from collections.abc import Callable
from typing import Union

# isort: off
from typing import Literal

# isort: on
from tvm_ffi import register_object

from tvm.ir import IRModule
from tvm.runtime import Object, Tensor
from tvm.target import Target

from .. import _ffi_api


@register_object("s_tir.meta_schedule.BuilderInput")
class BuilderInput(Object):
    """The builder's input.

    Parameters
    ----------
    mod : IRModule
        The IRModule to be built.
    target : Target
        The target to be built for.
    params: Optional[Dict[str, Tensor]]
        The parameters for Relax build module
    """

    mod: IRModule
    target: Target
    params: dict[str, Tensor] | None

    def __init__(
        self,
        mod: IRModule,
        target: Target,
        params: dict[str, Tensor] | None = None,
    ) -> None:
        """Constructor.

        Parameters
        ----------
        mod : IRModule
            The IRModule to be built.
        target : Target
            The target to be built for.
        params: Optional[Dict[str, Tensor]]
            The parameters for Relax build module
        """
        self.__init_handle_by_constructor__(
            _ffi_api.BuilderInput,  # type: ignore
            mod,
            target,
            params,
        )


@register_object("s_tir.meta_schedule.BuilderResult")
class BuilderResult(Object):
    """The builder's result.

    Parameters
    ----------
    artifact_path : Optional[str]
        The path to the artifact.
    error_msg : Optional[str]
        The error message.
    """

    artifact_path: str | None
    error_msg: str | None

    def __init__(
        self,
        artifact_path: str | None,
        error_msg: str | None,
    ) -> None:
        """Constructor.

        Parameters
        ----------
        artifact_path : Optional[str]
            The path to the artifact.
        error_msg : Optional[str]
            The error message.
        """
        self.__init_handle_by_constructor__(
            _ffi_api.BuilderResult,  # type: ignore
            artifact_path,
            error_msg,
        )


@register_object("s_tir.meta_schedule.Builder")
class Builder(Object):
    """The abstract builder interface."""

    BuilderType = Union["Builder", Literal["local"]]

    def build(self, build_inputs: list[BuilderInput]) -> list[BuilderResult]:
        """Build the given inputs.

        Parameters
        ----------
        build_inputs : List[BuilderInput]
            The inputs to be built.
        Returns
        -------
        build_results : List[BuilderResult]
            The results of building the given inputs.
        """
        return _ffi_api.BuilderBuild(self, build_inputs)  # type: ignore

    @staticmethod
    def create(
        kind: Literal["local"] = "local",
        *args,
        **kwargs,
    ) -> Builder:
        """Create a Builder.

        Parameters
        ----------
        kind : Literal["local"]
            The kind of the builder. For now, only "local" is supported.

        Returns
        -------
        builder : Builder
            The builder created.
        """
        from . import LocalBuilder

        if kind == "local":
            return LocalBuilder(*args, **kwargs)  # type: ignore
        raise ValueError(f"Unknown Builder: {kind}")


create = Builder.create


@register_object("s_tir.meta_schedule.PyBuilder")
class _PyBuilder(Builder):
    """
    A TVM object builder to support customization on the python side.
    This is NOT the user facing class for function overloading inheritance.

    See also: PyBuilder
    """

    def __init__(self, f_build: Callable | None = None):
        """Constructor."""

        self.__init_handle_by_constructor__(
            _ffi_api.BuilderPyBuilder,  # type: ignore
            f_build,
        )


class PyBuilder:
    """
    An abstract builder with customized build method on the python-side.
    This is the user facing class for function overloading inheritance.

    Note: @derived_object is required for proper usage of any inherited class.
    """

    _tvm_metadata = {"cls": _PyBuilder, "methods": ["build"]}

    def build(self, build_inputs: list[BuilderInput]) -> list[BuilderResult]:
        """Build the given inputs.

        Parameters
        ----------
        build_inputs : List[BuilderInput]
            The inputs to be built.
        Returns
        -------
        build_results : List[BuilderResult]
            The results of building the given inputs.
        """
        raise NotImplementedError
